---
title: "fig03"
author: "kaehwan"
date: "6/15/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
true <- TRUE
false <- FALSE
```

#### Load libraries

```{r}
source("configuration.R")
library(ggtext)
library(gridExtra)
library(scales)
```

#### Heatmap summary of spike-in samples

```{r}

res_summary <- read.table("../tables/metagenomics_spike_illumina_workflow_summary.txt",
                    header=TRUE, sep="\t", row.names=1) %>% 
  rownames_to_column("Workflow") %>% 
  rename(`Non-spike`=`Non.spike`, `10^1 - 10^2`=`X1.log`,
         `10^2 - 10^3`=`X2.log`, `10^3 - 10^4`=`X3.log`,
         `10^4 - 10^5`=`X4.log`)
res_summary$Workflow <- c("Standard GBS<br>isolation", 
                   "Mapping-based<br>GBS detection", 
                   "Assembly-based<br>Sequence typing", 
                   "Mapping-based<br>Sequence typing")

melt(res_summary) %>% 
        rename(`GBS spike level (cfu)`=variable, Value=value) %>%
        mutate(Workflow=factor(Workflow, levels=c("Standard GBS<br>isolation", 
                                                  "Mapping-based<br>GBS detection",
                                                  "Assembly-based<br>Sequence typing",
                                                  "Mapping-based<br>Sequence typing"))) %>%
        ggplot2::ggplot(ggplot2::aes(x=`GBS spike level (cfu)`, y=Workflow, fill=Value)) +
        ggplot2::geom_tile() +
        ggplot2::geom_text(ggplot2::aes(label = Value), size=6, fontface = "bold") +
        ggplot2::scale_fill_gradient(low = "white", high = "#1b98e0") +
        ggplot2::scale_y_discrete(limits=rev) +
        ggplot2::theme(axis.text.y=element_markdown(), 
                       axis.text.x=element_markdown(angle=0, vjust=0.5, hjust=0.5),
                       legend.position="none")



# Summary of benchmarking w/ GBS spike-in experiments ---------------------
meta <- read.table("../metadata/metagenomics_spike_metadata.txt",
                    header=TRUE, sep="\t") %>%
  rename(sample_id=Sample_ID, spike_level=GBS_spike_level, 
         SI=GBS_isolation, sMD=GBS_map_detected, sMT=GBS_map_seqTyped) %>%
  select(sample_id, spike_level, SI, sMD, sMT) %>%
  mutate(spike_level=recode(spike_level, '1-log'='10^1 - 10^2', 
                            '2-log'='10^2 - 10^3', '3-log'='10^3 - 10^4', 
                            '4-log'='10^4 - 10^5'))

meta %>%
        pivot_longer(-c('sample_id', 'spike_level'), names_to='workflows', values_to='values') %>% 
        mutate(spike_level=factor(spike_level, levels=c('Non-spike', 
                                                        '10^1 - 10^2', 
                                                        '10^2 - 10^3', 
                                                        '10^3 - 10^4', 
                                                        '10^4 - 10^5')), 
               workflows=factor(workflows, levels=c('SI', 'sMD', 'sMT')),
               values=factor(values, levels=c('Yes', 'No'))) %>% 
        ggplot2::ggplot(ggplot2::aes(x=workflows, y=sample_id, fill=values)) +
        ggplot2::geom_tile(color='black', size=0.35) +
        ggplot2::facet_grid(.~ spike_level) +
        ggplot2::coord_fixed() +
        ggplot2::scale_fill_manual(name='', values=c('brown', 'lightgrey')) +
        ggplot2::labs(x='', y='Samples', title="GBS spike level (cfu)") +
        ggplot2::theme(panel.border=ggplot2::element_blank(),
                       axis.ticks=ggplot2::element_blank(),
                       plot.title=ggplot2::element_text(size=22),
                       axis.text.x=ggplot2::element_text(angle=0),
                       strip.text.x=element_markdown(),
                       strip.background=ggplot2::element_blank())
ggplot2::ggsave('../output_figures/fig03_b_illumina_spike_result_summary.svg', 
       width=14, height=8)

```


#### Checking assembled genome completeness of real-world and spike-in samples

```{r}
# import real-world samples metadata
meta <- read.table('../metadata/metagenomics_control_metadata.txt', 
                   header=TRUE, row.names=4, sep="\t") %>% 
        rownames_to_column("Library") %>% 
        select(Library, GBS_isolated, GBS_map_detected, GBS_map_seqTyped)
# import spike-in samples metadata
meta.spike <- read.table('../metadata/metagenomics_spike_metadata.txt',
                         header=TRUE, row.names=4, sep="\t") %>% 
        rownames_to_column("Library") %>% 
        select(Library, GBS_spike_level, GBS_isolation, GBS_map_detected, GBS_map_seqTyped)
# import real-world metaphlan4 relative abundance        
dat <- read.table("../output_tables/metagenomics.metaphlan4.control.filtered.s", 
                  header=TRUE, row.names=1) %>% 
        mutate(across(where(is.numeric), round, digits=4))
dat[dat < 0.1] <- 0
dat <- dat[rowSums(dat)>0, ] %>% 
        t() %>% 
        as.data.frame() %>% 
        select("s__Streptococcus_agalactiae") %>% 
        rename(`Streptococcus agalactiae`=s__Streptococcus_agalactiae) %>% 
        rownames_to_column("Library") 
# import spike-in metaphlan4 relative abundance
dat.spike <- read.table("../output_tables/metagenomics.metaphlan4.spike.filtered.s", 
                        header=TRUE, row.names=1) %>% 
        mutate(across(where(is.numeric), round, digits=4))
dat.spike[dat.spike < 0.1] <- 0
dat.spike <- dat.spike[rowSums(dat.spike)>0, ] %>% 
        t() %>% 
        as.data.frame() %>% 
        select("s__Streptococcus_agalactiae") %>% 
        rename(`Streptococcus agalactiae`=s__Streptococcus_agalactiae) %>% 
        rownames_to_column("Library")
# import QUAST result of natural real-world fish samples
quast.result <- read.table("../tables/quast/metagenomics_control_gbs_quast_report.tsv", 
                           header=TRUE, row.names=1, sep="\t") %>% 
        t() %>% 
        as.data.frame() %>% 
        rownames_to_column("Library") %>% 
        select(Library, `Total length`, N50) %>% 
        mutate(Library=gsub(".scaffolds", "", Library)) %>% 
        mutate(Library=gsub(".spades.step", "", Library)) %>% 
        mutate(Library=gsub(".final_assembly", "", Library))
# import QUAST result of spike-in fish samples
quast.spike <- read.table("../tables/quast/metagenomics_spike_gbs_quast_report.tsv",
                          header=TRUE, row.names=1, sep="\t") %>% 
        t() %>% 
        as.data.frame() %>% 
        rownames_to_column("Library") %>% 
        select(Library, `Total length`, N50) %>% 
        mutate(Library=str_replace_all(Library, ".spades.step", ""))
# # import real-world CheckM result
# checkm <- read.csv("../tables/checkm.mapAssembly.control.csv",
#                           header=TRUE, row.names=2) %>%
#         rownames_to_column("Library") %>% 
#         rename(Quality=QUAL_y) %>% 
#         select(Library, Experiment, Completeness, Contamination, Quality) %>% 
#         mutate(Quality=factor(Quality, levels=c("HIGH", "MODERATE", 
#                                                 "LOW", "INCOMPLETE")))
# # import spike-in CheckM result
# checkm.spike <- read.csv('../tables/checkm.mapAssembly.spike.csv',
#                          header=TRUE, row.names=2) %>% 
#         rownames_to_column("Library") %>% 
#         rename(Quality=QUAL_Y) %>% 
#         select(Library, Experiment, Completeness, Contamination, Quality) %>% 
#         mutate(Quality=factor(Quality, levels=c("HIGH", "MODERATE",
#                                                 "LOW", "INCOMPLETE")))
# inner_join all real-world data: metadata, metaphlan4 relative abundance, QUAST, CheckM
df <- inner_join(meta, quast.result, by="Library") %>% 
        merge(., dat, by="Library") %>% 
        # merge(., checkm, by="Library") %>%
        rename(`GBS detected`=GBS_map_detected, `Sequence typed`=GBS_map_seqTyped) %>% 
        mutate(`GBS detected`=factor(`GBS detected`, levels=c("Yes", "No"))) %>% 
        mutate(`Sequence typed`=factor(`Sequence typed`, levels=c("Yes", "No"))) %>% 
        mutate(`GBS isolation`=if_else(`GBS_isolated`=="Yes", "Isolated", "Not isolated")) %>% 
        mutate(`Metagenomics workflow`=if_else(`GBS detected`=="No", 
                                               "No GBS detected", 
                                               if_else(`GBS detected`=="Yes"&`Sequence typed`=="No", 
                                                       "GBS detected", 
                                                       "GBS detected & ST typed"))) %>% 
        mutate(`Metagenomics workflow`=factor(`Metagenomics workflow`, 
                                              levels=c("No GBS detected", 
                                                       "GBS detected", 
                                                       "GBS detected & ST typed"))) %>% 
        mutate(`Total length`=if_else(`GBS detected`=='No', 0, `Total length`)) %>% 
        select(-c(GBS_isolated, `GBS detected`, `Sequence typed`))
# inner_join all spike-in data: metadata, metaphlan4 relative abundance, QUAST, CheckM
df.spike <- inner_join(meta.spike, quast.spike, by="Library") %>% 
        merge(., dat.spike, by="Library") %>% 
        # merge(., checkm.spike, by="Library") %>%
        rename(`GBS detected`=GBS_map_detected, `Sequence typed`=GBS_map_seqTyped) %>% 
        mutate(`GBS detected`=factor(`GBS detected`, levels=c("Yes", "No"))) %>% 
        mutate(`Sequence typed`=factor(`Sequence typed`, levels=c("Yes", "No"))) %>% 
        mutate(`GBS isolation`=if_else(`GBS_isolation`=="Yes", "Isolated", "Not isolated")) %>% 
        mutate(`Metagenomics workflow`=if_else(`GBS detected`=="No", 
                                               "No GBS detected", 
                                               if_else(`GBS detected`=="Yes"&`Sequence typed`=="No", 
                                                       "GBS detected", 
                                                       "GBS detected & ST typed"))) %>% 
        mutate(`Metagenomics workflow`=factor(`Metagenomics workflow`, 
                                              levels=c("No GBS detected", 
                                                       "GBS detected", 
                                                       "GBS detected & ST typed"))) %>%
        mutate(GBS_spike_level=factor(GBS_spike_level, levels=c('Non-spike', 
                                                                '1-log', 
                                                                '2-log', 
                                                                '3-log', 
                                                                '4-log'))) %>% 
        mutate(`Total length`=if_else(`GBS detected`=='No', 0, `Total length`)) %>% 
        select(-c(GBS_isolation, `GBS detected`, `Sequence typed`))


# Analysis for artificial spike-in samples
df.spike %>% 
        mutate(`GBS_spike_level`=recode(GBS_spike_level, 
                                        '1-log'='10^1 - 10^2', '2-log'='10^2 - 10^3', 
                                        '3-log'='10^3 - 10^4', '4-log'='10^4 - 10^5')) %>%
        mutate(`Streptococcus agalactiae`=if_else(`Streptococcus agalactiae`==0,
                                                  `Streptococcus agalactiae`+0.01,
                                                  `Streptococcus agalactiae`)) %>%
        ggplot2::ggplot(ggplot2::aes(x=`Total length`/2.21e6*100, y=`Streptococcus agalactiae`)) + 
        ggplot2::geom_jitter(ggplot2::aes(shape=`Metagenomics workflow`, fill=`GBS_spike_level`), 
                    stroke=1, size=12, width=1.5, height=0.2) + 
        ggplot2::scale_y_log10(breaks=c(0.01, 0.1,1,10,100), labels=c(0.01,0.1,1,10,100)) +
        ggplot2::scale_fill_manual(name="Metagenomics<br>analysis", 
                          values=c("#e7298a", "#D95F02", "#e6ab02", "#7570B3", "#1B9E77")) +
        ggplot2::scale_shape_manual(name='GBS spike level (cfu)', values=c(21,22,23,24,25)) +
        ggplot2::guides(fill=ggplot2::guide_legend(override.aes=list(size=6))) +
        ggplot2::annotation_logticks(sides="l") +
        ggplot2::labs(y="GBS rel. abund. (%)", x="Assembled genome completeness (%)") +
        ggplot2::theme(legend.title=element_markdown(size=20),
              legend.text=element_markdown())
ggplot2::ggsave('../output_figures/fig03_c_illumina_spike_assembly_analysis.svg', 
       width =12, height = 8)

# Analysis for natural real-world samples
df %>% 
        mutate(`GBS isolation`=recode(`GBS isolation`, 
                                      'Isolated'='Positive', 
                                      'Not isolated'='Negative')) %>% 
        mutate(`Streptococcus agalactiae`=if_else(`Streptococcus agalactiae`==0,
                                                  `Streptococcus agalactiae`+0.01,
                                                  `Streptococcus agalactiae`)) %>%
        ggplot2::ggplot(ggplot2::aes(x=`Total length`/2.21e6*100, y=`Streptococcus agalactiae`)) + 
        ggplot2::geom_jitter(ggplot2::aes(fill=`Metagenomics workflow`, shape=`GBS isolation`), 
                   stroke=1, size=12, width=1.5, height=0.2) + 
        ggplot2::scale_y_log10(breaks=c(0,0.01,0.1,1,10,100), labels=c(0,0.01,0.1,1,10,100)) +
        ggplot2::scale_fill_manual(name="Metagenomics<br>Analysis", 
                          values=c("#7570B3", "#D95F02", "#1B9E77")) +
        ggplot2::scale_shape_manual(name='GBS Isolation', values=c(21,22,23,24,25)) +
        ggplot2::guides(fill=ggplot2::guide_legend(override.aes=list(size=5))) +
        ggplot2::annotation_logticks(sides="l") +
        ggplot2::labs(y="GBS relative abundance (%)", x="Assembled genome completeness (%)") +
        ggplot2::theme(legend.title=element_markdown(size=20))
ggplot2::ggsave('../output_figures/fig03_d_illumina_control_assembly_qc.svg', 
       width =12, height = 8)
```

#### Plotting SNP frequencies of recovered GBS genome from real-world fish samples M0045

```{r}
path <- "/Users/simkaehwan/Documents/Bioinfo/metaG/tables/vcf/"

vcf <- paste0(path, list.files(path, pattern=".vcf")) %>% 
  map_df(~fread(., skip="#CHROM", data.table=F)) %>% 
  pivot_longer(cols=starts_with("/mnt/ab2/common/illumina/metaG_nature/align2ref2/"), 
               names_to="Sample", 
               names_prefix="/mnt/ab2/common/illumina/metaG_nature/align2ref2/", 
               values_to="Record", 
               values_drop_na=TRUE) %>% 
  rename(CHROM="#CHROM") %>% 
  select(CHROM, POS, REF, ALT, FORMAT, Sample, Record) %>% 
  separate(col=Record, into=c("GT", "PL", "DP", "SP", "ADF", "ADR", "AD"), sep=":") %>% 
  separate(col=ALT, into=c("ALT1", "ALT2", "ALT3"), sep=",") %>% 
  separate(col=AD, into=c("REF_count", "ALT1_count", "ALT2_count", "ALT3_count"), sep=",") %>% 
  mutate(Sample=gsub(".bam", "", Sample),
         ALT1_count=coalesce(as.numeric(ALT1_count), 0),
         ALT2_count=coalesce(as.numeric(ALT2_count), 0),
         ALT3_count=coalesce(as.numeric(ALT3_count), 0),
         AAF=(ALT1_count+ALT2_count+ALT3_count)/as.numeric(DP)) %>% 
  mutate(Ref="RefDB")

vcf %>% 
  filter(Sample=="M0045") %>% 
  ggplot(aes(x=AAF, fill=Ref, color=Ref)) +
  geom_histogram(binwidth=0.1, alpha=0.5, position="identity", color="darkblue", fill="lightblue") +
  facet_wrap(~Sample, scale="free") +
  scale_x_continuous(limits=c(0, 1.1), breaks=c(0.0, 0.5, 1.0)) +
  theme(strip.background=element_rect(fill="white", color="white"))
ggsave('../output_figures/fig03_e_illumina_control_snp_qc.svg', width=8, height=8)
```

#### Session info

```{r}
sessionInfo()
```