---
title: "supfig02"
author: "kaehwan"
date: "7/26/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
true <- TRUE
false <- FALSE
```

```{r}
library(data.table)
library(tidyverse)
```

#### Plotting SNP frequencies of recovered GBS genomes from fish samples naturally contaminated with GBS

```{r}
path <- "/Users/simkaehwan/Documents/Bioinfo/metaG/tables/vcf/"

vcf <- paste0(path, list.files(path, pattern=".vcf")) %>% 
  map_df(~fread(., skip="#CHROM", data.table=F)) %>% 
  pivot_longer(cols=starts_with("/mnt/ab2/common/illumina/metaG_nature/align2ref2/"), 
               names_to="Sample", 
               names_prefix="/mnt/ab2/common/illumina/metaG_nature/align2ref2/", 
               values_to="Record", 
               values_drop_na=TRUE) %>% 
  rename(CHROM="#CHROM") %>% 
  select(CHROM, POS, REF, ALT, FORMAT, Sample, Record) %>% 
  separate(col=Record, into=c("GT", "PL", "DP", "SP", "ADF", "ADR", "AD"), sep=":") %>% 
  separate(col=ALT, into=c("ALT1", "ALT2", "ALT3"), sep=",") %>% 
  separate(col=AD, into=c("REF_count", "ALT1_count", "ALT2_count", "ALT3_count"), sep=",") %>% 
  mutate(Sample=gsub(".bam", "", Sample),
         ALT1_count=coalesce(as.numeric(ALT1_count), 0),
         ALT2_count=coalesce(as.numeric(ALT2_count), 0),
         ALT3_count=coalesce(as.numeric(ALT3_count), 0),
         AAF=(ALT1_count+ALT2_count+ALT3_count)/as.numeric(DP)) %>% 
  mutate(Ref="RefDB")


vcf %>% 
  filter(Sample!="M0045") %>% 
  ggplot(aes(x=AAF, fill=Ref, color=Ref)) +
  geom_histogram(binwidth=0.1, alpha=0.5, position="identity", color="darkblue", fill="lightblue") +
  facet_wrap(~Sample, scale="free") +
  scale_x_continuous(limits=c(0, 1.1), breaks=c(0.0, 0.5, 1.0)) +
  theme(strip.background=element_rect(fill="white", color="white"))
ggsave('../output_figures/supfig02_illumina_control_snp_qc.svg', width=12, height=8)
```

```{r}
sessionInfo()
```


